Refactor and improve Previewgenerator

szaimen opened the pull request:
See: https://github.com/nextcloud/vm/issues/1001

steps:
- [x] change preview settings
- [x] let us choose the preview-filetypes
- [x] let us choose if we want to install imagick
- [x] run the previewgenerator at 4 am
- [x] define a nextcloud-user for which the cronjob ist running
- [x] be able to run the script a second time

szaimen commented:
@enoch85 I would like to get some early feedback on this. Is my first PR here on GitHub. :)

enoch85 commented:
Thanks for your PR! <3

You need to fix the checks here: https://travis-ci.org/nextcloud/vm/builds/630331851?utm_source=github_status&utm_medium=notification

I will have a closer look when I'm back from vacation.

szaimen commented:
@enoch85 Travis should be fixed by now.

Actually I don't know how I could do this:
- let us choose the preview-filetypes
- define a nextcloud-user for which the cronjob ist running
Maybe do you have some example code which I can look at?

But this could also be something for another PR.

Of course this can wait till after your vacation.

enoch85 commented on apps/previewgenerator.sh:
```diff
@@ -60,23 +74,32 @@ then
     
-    # Pre generate everything
-    occ_command preview:generate-all
+    Please note: If you choose to not doing it now, then you have to do it manually before the cronjob at the next day at 4 am can continue.
```
Remove this, there shouldn't be an option *not* to run the pre-generate command.

szaimen replied:
Done.

enoch85 commented on apps/previewgenerator.sh:
```diff
@@ -16,18 +16,32 @@ debug_mode
     sleep 1
 fi
 
+msg_box "In the next step you can choose to install a package called imagick to speed up the generation of previews and get support for many filetypes. 
```
I'm not sure but doesn't the previewgenerator depend on imagick? In other words, if not installed previewgenerator will not function?

szaimen replied:
No, I am currently running the preview:generate-all command on my instance without imagick installed and it works nicely.

szaimen commented:
@enoch85 It is ready for review again. 
I know, that there are some mistakes now, but I don't know how to fix them. 
Maybe you could help me out after your vacation?

enoch85 commented:
There are more errors that needs to be fixed: https://travis-ci.org/nextcloud/vm/builds/630885829?utm_source=github_status&utm_medium=notification

Please have a look.

enoch85 commented:
Also, problem with *not* installing Imagick here is that if not all filetypes are supported, then I'm sure we'll get a bug report saying: Previewrgenerator for XYZ is not working" because the user didn't enable Imagick and expect everything to work anyway. Most users don't read documentation, or even the text in the `msg_box`, they just expect everything to work.

As this adds more complexity it's very important to be more clear on exactly what happens and why.

Example, if you come from Windows and never saw a console box before, you would be very unfamiliar with everything, also take in to consideration that you don't read the informatic boxes - what would the end result be in 6 months from now when you decide to add support for SVG (since you got better in Nextcloud and  now knows what the `config.php` file does) and you forgot to add Imagick, or didn't add it because "it wasn't forced / needed) at the time? Answer: a bug report, a frustrated user, and a worse VM experience because the user will think it was the VMs fault, and not her/his own decision 6 months ago.

What I'm trying to say, you have to think of all scenarios when designing this stuff. :)

szaimen commented:
> There are more errors that needs to be fixed: https://travis-ci.org/nextcloud/vm/builds/630885829?utm_source=github_status&utm_medium=notification
> 
> Please have a look.

Actually I am not an experienced programmer and currently don't know, how to fix Travis.

What I have done is copying the part from [here](https://github.com/nextcloud/vm/blob/9c73939ac395f0091e7ca33c73311eb869dfffcf/lets-encrypt/activate-ssl.sh#L63) downwards and replaced everything with "domain" in it through "user" and pasted it [here](https://github.com/nextcloud/vm/pull/1002/commits/d5aa95e577a694a98c5abc5bf2f3ea0ded9179c1).
What I wanted to do there is letting the user enter a nextcloud-user for which the cronjob and the preview-generation at the end is running.
But that is probably not what I have done with the code I have added.

So could you maybe give me some code examples on how I could do it right?
Would be really appreciated!

szaimen commented:
> Also, problem with _not_ installing Imagick here is that if not all filetypes are supported, then I'm sure we'll get a bug report saying: Previewrgenerator for XYZ is not working" because the user didn't enable Imagick and expect everything to work anyway. Most users don't read documentation, or even the text in the `msg_box`, they just expect everything to work.
> 
> As this adds more complexity it's very important to be more clear on exactly what happens and why.
> 
> Example, if you come from Windows and never saw a console box before, you would be very unfamiliar with everything, also take in to consideration that you don't read the informatic boxes - what would the end result be in 6 months from now when you decide to add support for SVG (since you got better in Nextcloud and now knows what the `config.php` file does) and you forgot to add Imagick, or didn't add it because "it wasn't forced / needed" at the time? Answer: a bug report, a frustrated user, and a worse VM experience because the user will think it was the VMs fault, and not her/his own decision 6 months ago.
> 
> What I'm trying to say, you have to think of all scenarios when designing this stuff. :)

I will try to find out which formats are not supported without imagick and then edit my PR to just allow the formats that are available depending on if you have chosen on installing imagick or not.

szaimen commented:
So, I tested the different file formats out:
The following is the case (I first enabled all formats):
- if you just install ffmpeg these formats work:
PNG
JPEG
GIF
BMP
MP3
TXT
Markdown
Movie

-if you then install imagick these additional formats work:
Photoshop
SVG
TIFF

-I couln't make the following formats work (also if imagick and libreoffice is installed):
HEIC
MSOffice2003
MSOffice2007
MSOfficeDoc
OpenDocument
PDF
Postscript
StarOffice
Font
XBitmap
Illustrator

So either they are no longer supported, or something is broken in the previewgenerator or somewhere else.
Actually when generating the previews, I always got this error for office documents (when imagick (with and without libreoffice) is installed):
![image](https://user-images.githubusercontent.com/42591237/71604488-a566f980-2b62-11ea-97d5-00c5078a4921.png)
That could be the reason why the previews for these formats don't get generated.

So I would then change my script to represent the working files.

enoch85 commented:
> What I wanted to do there is letting the user enter a nextcloud-user for which the cronjob and the preview-generation at the end of the script is running.

All previews should be generated and that's done with the existing `www-data`. I don't see the point with doing it with a second Nextcloud user. Please enlighten me.

szaimen commented:
> > What I wanted to do there is letting the user enter a nextcloud-user for which the cronjob and the preview-generation at the end of the script is running.
> 
> All previews should be generated and that's done with the existing `www-data`. I don't see the point with doing it with a second Nextcloud user. Please enlighten me.

Sry for not beeing clear.
I didn't mean a system-user but a nextcloud-user (from the instance).
Because in the preview generation command you can define a user for which you want to run the command and cronjob:
https://github.com/rullzer/PreviewGenerator#previewgenerate-all-user-id

This can especially be useful, if you use groupfolders or the external storage app to mount folders to many nextcloud-users with a lot of files in it. Because then you mount all folders for one puppet-user and the folders then get scanned only one time and not for each user again which isn't really efficient.

But if you don't need to define a nextcloud-user for the cronjob, you can just skip this step.

szaimen commented:
> > > What I wanted to do there is letting the user enter a nextcloud-user for which the cronjob and the preview-generation at the end of the script is running.
> > 
> > 
> > All previews should be generated and that's done with the existing `www-data`. I don't see the point with doing it with a second Nextcloud user. Please enlighten me.
> 
> Sry for not beeing clear.
> I didn't mean a system-user but a nextcloud-user (from the instance).
> Because in the preview generation command you can define a user for which you want to run the command and cronjob:
> https://github.com/rullzer/PreviewGenerator#previewgenerate-all-user-id
> 
> This can especially be useful, if you use groupfolders or the external storage app to mount a few folders to many nextcloud-users with a lot of files in it. Because then you mount all folders to one puppet-user and the cronjob then runs for this single user and not for all users, which would result in these folders getting scanned a lot of times each time you run the cronjob.
> 
> So would be really nice to get this into the script, because if you don't need to define a nextcloud-user for the cronjob, you can just skip this last step.

I think I found a way to doing it right in my last commit.

enoch85 commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,218 @@ debug_mode
     
-    # Add crotab
-    crontab -u www-data -l | { cat; echo "@daily php -f $NCPATH/occ preview:pre-generate >> /var/log/previewgenerator.log"; } | crontab -u www-data -
+    msg_box "In the last step you can define a nextcloud-user for which you want to run the preview-generation. "
```
Please remove this option. If the user makes a typo, the pre-generate will never happen. Also, as it's a cronjob, it will run in the background and no one will notice as it's in the middle of the night.

If you *really* think this option is useful and want to add this feature, please make `$nextcloud_user` `www-data` per default, and rename the variable to `cronjob_user`. So that if nothing is chosen, use www-data.

Something like: "The default user for generating previews are www-data. There's no reason to change that expect in special edge cases. Please only change this if you know what you are doing!"

Adding complexity = more likely to get bugs. Keep it simple. :)

enoch85 replied:
Also, if you go for the second option, add a check with `occ_command user:list | grep $user` to check if the user exists. If not, then ask for input again.

Thanks.

szaimen replied:
The thing is, that I don't want to change the cronjob-user but define a nextcloud-user for which I want to run the command as I explained here: https://github.com/nextcloud/vm/pull/1002#issuecomment-569841096
But I added the check. Thanks for the advice.

enoch85 replied:
It's not what you want, it's what everyone wants. :) This VM is downloaded several thousand times. Having a setting that can be confusing (e.g. why even run with a certain user? Not everyone have the same setup as you.) is asking for issue reports when users say: "I installed preview generator, but it doesn't work".

Please change so that `www-data` is default, add text to explain what happens if you change it, like you explained for me. www-data is default for every user, having a single user is *very much an edge case* which involves group folders and so on.

szaimen replied:
actually I don't know what you mean. The system user is actually set to www-data
and if not set the preview generation runs for all nextcloud-users as default.

look at line 243:
`crontab -u www-data -l | { cat; echo "0 4 * * * php -f $NCPATH/occ preview:pre-generate $nextcloud_user >> /var/log/previewgenerator.log"; } | crontab -u www-data -`

so if the $nextcloud_user variable is empty, nothing gets added. So then it runs for all nextcloud-users.

szaimen replied:
I have tried to improve the explanation of this a bit. and make sure that it runs for all nextcloud-users, if no user is set up

enoch85 commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,218 @@ debug_mode
+
     # Pre generate everything
-    occ_command preview:generate-all
+    occ_command preview:generate-all $nextcloud_user
```
We want to pre-generate for all users, not only a single user.

Please remove the specific user.

szaimen replied:
If a user is not defined, the previews get generated for all users.

szaimen replied:
If it is better for you, I could move this part into the if-else-then and completely separate it, depending on if you choose to select a nextcloud user or not.

szaimen replied:
I've done this now.
Maybe it is a bit more clear now.

enoch85 commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,218 @@ debug_mode
+msg_box "In the next step you can choose to install a package called imagick to speed up the generation of previews and get support for more filetypes. 
 
-# Run the first preview generation and add crontab
-if [ -d "$NC_APPS_PATH/previewgenerator" ]
```
Why was this check removed?

szaimen replied:
I copied this to line 200, because I think there is the only command for which you really need the previewgenerator enabled.
If you want, I can reverse this change.

szaimen replied:
I moved it back to the start of the script.
So: Done.

szaimen commented:
My PR is ready for review again. Travis is finally fixed.

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,218 @@ debug_mode
+    do
+        case $choice in
+            PNG)
+                clear
```
I removed all except for the first one.

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,218 @@ debug_mode
+    do
+        case $choice in
+            PNG)
+                clear
```
I removed all except for the first one.

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,209 @@ debug_mode
+    then
+        nextcloud_user=$(whiptail --inputbox "Enter the nextcloud-user for which you want to run the preview-generation" 10 30 3>&1 1>&2 2>&3)
+        export nextcloud_user
+        if [[ "yes" == $(ask_yes_or_no "Is this correct? $nextcloud_user") ]]
```
@enoch85 btw: do we need this check? (Because what should happen if you press no?)

enoch85 replied:
Better safe than sorry. :)

enoch85 replied:
I will improve everything once I have time. Just commenting the obvious now.

enoch85 commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,209 @@ debug_mode
+#occ_command config:system:set preview_libreoffice_path --value="/usr/bin/libreoffice"
+#occ_command config:system:set enabledPreviewProviders 4 --value="OC\\Preview\\XBitmap"
+#occ_command config:system:set enabledPreviewProviders 8 --value="OC\\Preview\\Illustrator"
+#occ_command config:system:set enabledPreviewProviders 10 --value="OC\\Preview\\MSOffice2003"
```
Doesn't MSOffice and PDF work if you install the libreoffice packages (which you now seems to have removed)?

szaimen replied:
No, I have tested this out and it hasn't worked also when libreoffice was installed
but look here for details:
https://github.com/nextcloud/vm/pull/1002#issuecomment-569828067

enoch85 replied:
Aah ok, thanks for confirming. 

Maybe remove those lines and refer to the Nextcloud documentation instead to keep it clean.

szaimen replied:
I removed them and added the comment to the specific section (currently line 79 and 154)

enoch85 commented on apps/previewgenerator.sh:
```diff
@@ -110,6 +112,10 @@ then
 
 else
+    # remove imagick
+    apt purge php-imagick -y
```
This is not needed as it's not installed in the VM. Good thinking though.

szaimen replied:
if you run the script a second time it could happen though.

enoch85 replied:
Then please check if the package exists first. Thanks.

szaimen replied:
Have added the check.
So should be done.

enoch85 commented:
I will take some time to test this now. If all is good I will merge this today if you don't have any other additions?

enoch85 commented:
BUG: sed: -e expression #1, char 2: extra characters after command

szaimen commented:
> I will take some time to test this now. If all is good I will merge this today if you don't have any other additions?

No, I am quiet satisfied with the options and flow the new script brings with it.
So go ahead and merge it when you are satisfied too and if there are no bugs left. 👍

enoch85 commented:
Have you tested this yourself? I'm asking because the whiptail doesn't seem to work, and never was.

What do you say about enabling all previews instead of choosing? I mean, why wouldn't you want to enable everything?

enoch85 commented:
There are multiple bugs, I don't have more time now. 

Please have a look yourself.

szaimen commented:
> Have you tested this yourself? I'm asking because the whiptail doesn't seem to work, and never was.

whiptails are working now.

> What do you say about enabling all previews instead of choosing? I mean, why wouldn't you want to enable everything?

e.g. I don't like txt previews. They don't make much sense imo.
The same for markdown and mp3
... so having the choice is better imho.

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,258 @@ debug_mode
+        
+        # reset the cronjob
+        print_text_in_color "$ICyan" "Resetting the cronjob for the preview-generation"
+        crontab -u www-data -l | grep -v 'preview:generate-all'  | crontab -u www-data -
```
@enoch85 this doesn't remove anything in the moment.
do you have another idea what to change to make it work?

enoch85 replied:
When I tested it, it removed everything. Using `sed` isn't a clean solution imo.

szaimen replied:
Hm...

enoch85 replied:
It works as expected.

szaimen replied:
Okay

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,258 @@ debug_mode
+
+msg_box "In the last step you can define a specific Nextcloud user for which will be the user that runs the preview-generation. 
+
+The default behavoiur (just hit [ENTER]) is to run with the system user 'www-data' which will generate previews for all users. 
```
hitting enter deosn't work here.
So I could remove that line?

enoch85 replied:
No please leave it. I'll test everything more later.

enoch85 replied:
Enter choose the default - `www-data`.

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,258 @@ debug_mode
+    PREVIEW_USER=$(whiptail --inputbox "Enter the Nextcloud user for which you want to run the preview-generation" "$WT_HEIGHT" "$WT_WIDTH" 3>&1 1>&2 2>&3)
+    if ! occ_command user:list | grep "$PREVIEW_USER" | awk '{print $3}'
+        then
+            msg_box "It seems like the user you entered ($PREVIEW_USER) doesn't exist, please try again."
```
the check somehow doesn't get triggered. 
do you have another idea how to do it?

enoch85 replied:
Yeah, that's one of the bugs. I'll think about it. If you remove `awk print` you get another error.

enoch85 replied:
The most "secure" option would be to use `read` as we do in the other scrips.

enoch85 replied:
This is fixed.

enoch85 commented:
OK, I'm happy about this now in general. Maybe some smaller fixes, like wording and similar, but code wise I think it looks good. :+1:

szaimen commented:
> OK, I'm happy about this now in general. Maybe some smaller fixes, like wording and similar, but code wise I think it looks good.

Very Nice! 🎉

enoch85 commented:
Nice work @szaimen! :+1: :rocket: 

Please test this thoroughly, if you don't find anything, I'll merge tomorrow.

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,256 @@ debug_mode
+
+The default behaviour (just hit [ENTER]) is to run with the system user 'www-data' which will generate previews for all users. 
+
+If you on the other hand choose to use a specific user, previews will ONLY be generated for that specific user."
```
I still think it is a bit confusing to mention the www-data user here since it is just important to run it for all nextcloud-users or only for a single one imho.
So maybe: 
`The befault behaviour (just hit [ENTER]) ist to generate previews for all users.`

szaimen replied:
We could also think here about adding a bit more Nextcloud user 
So that it is really clear that we mean Nextcloud users here (is just mentioned at the beginning)

szaimen commented:
> Nice work @szaimen! 👍 🚀
> 
> Please test this thoroughly, if you don't find anything, I'll merge tomorrow.

Thank you @enoch85!

Everything seems to work in my testing! So ready to merge imo. 🚀
But before, we could think one last time about this topic:
https://github.com/nextcloud/vm/pull/1002#discussion_r362350088

szaimen commented:
Maybe we could also mention somewhere that the script can run more than once to change some settings  later on?

szaimen commented on apps/previewgenerator.sh:
```diff
@@ -16,67 +16,256 @@ debug_mode
-if [[ "no" == $(ask_yes_or_no "Do you still want to continue?") ]]
+# Encryption may not be enabled
+if [ -n "$(occ_command app:list | grep encryption | awk '{print $3}')" ]
 then
```
Is this working for you?
I haven't tested it myself.
And what about end to end encryption?
Shouldn't that get checked here, too?

szaimen replied:
I guess my mistake. Should work with both, since they both have "encryption" in their name.

enoch85 commented:
It works as expected, merging!
